scheduler/scd_utils.py

```diff
@@ -0,0 +1,239 @@
+
+            scd_lines.append({"timestamp" : epoch_milliseconds,
+                                "time" : time,
+                                "duration" : str(duration),
+                                "prio" : str(prio),
+                                "experiment" : experiment})
+
+        return scd_lines
+
+    def fmt_line(self, line_dict):
+        """Formats a dictionary with line info into a text line for file.
+
+        Args:
+            line_dict (TYPE): A dict that holds all the line info.
+
+        Returns:
+            TYPE: Formatted string.
+        """
+        line_str = self.line_fmt.format(datetime=line_dict["time"].strftime(self.scd_dt_fmt),
+                                        prio=line_dict["prio"],
+                                        experiment=line_dict["experiment"],
+                                        duration=line_dict["duration"])
+        return line_str
+
+    def write_scd(self, scd_lines):
+        """Creates SCD text lines and writes to file. Creates a backup of the old file before
+        writing.
+
+        Args:
+            scd_lines (TYPE): A list dicts that contain the line info.
+        """
+        text_lines = [self.fmt_line(x) for x in scd_lines]
+
+        shutil.copy(self.scd_filename, self.scd_filename+".bak")
+
+        with open(self.scd_filename, 'w') as f:
+            for line in text_lines:
+                f.write("{}\n".format(line))
+
+
+    def add_line(self, mmddyy, hhmm, prio, experiment, duration='-'):
+        """Adds a new line to the SCD.
+
+        Args:
+            mmddyy (TYPE): month/day/year string.
+            hhmm (TYPE): hour/minute string.
+            prio (TYPE): priority value.
+            experiment (TYPE): The experiment to run.
+            duration (str, optional): an optional duration to run for.
+
+        Raises:
+            ValueError: If line parameters are invalid or if line is a duplicate.
+        """
+
+        try:
+            # create datetime from args to see if valid
+            time = dt.datetime.strptime(mmddyy + " " + hhmm, self.scd_dt_fmt)
+        except:
+            raise ValueError("Can not create datetime from supplied formats")
+
+        if not (0 <= prio <= 20):
+            raise ValueError("Priority is out of bounds. 0 <= prio <= 20.")
+
+
+        scd_lines = self.read_scd()
+
+        epoch = dt.datetime.utcfromtimestamp(0)
+        epoch_milliseconds = int((time - epoch).total_seconds() * 1000)
+
+        new_line = {"timestamp" : epoch_milliseconds,
+                    "time" : time,
+                    "duration" : str(duration),
+                    "prio" : str(prio),
+                    "experiment" : experiment}
+
+        if new_line in scd_lines:
+            raise ValueError("Line is a duplicate of an existing line")
+
+        scd_lines.append(new_line)
+
+        # sort first by timestamp, then by duration with default duration first, then by priority.
```

Comment:
I think sorting should be by timestamp first, then priority, then lastly by duration (with non-default duration being first). Default duration would run after the non-default is over.